src/numa.h

```diff
@@ -94,21 +83,219 @@ inline CpuIndex get_hardware_concurrency() {
+        return true;
+
+    auto [status, groupAffinity] = get_process_group_affinity();
+
+    // If GroupCount > 1 then we know old API was never used and we can stick
+    // to the new API safely.
+    if (status != 0 && groupAffinity.size() > 1)
+        return false;
+
+    return true;
+};
+
+// On Windows there are two ways to set affinity, and therefore 2 ways to get it.
+// These are not consistent, so we have to check both.
+// In some cases it is actually not possible to determine affinity.
+// For example when two different threads have affinity on different processor groups,
+// set using SetThreadAffinityMask, we can't retrieve the actual affinities.
+// From documentation on GetProcessAffinityMask:
+//     > If the calling process contains threads in multiple groups,
+//     > the function returns zero for both affinity masks.
+// In such cases we just give up and assume we have affinity for all processors.
+// nullopt means no affinity is set, that is, all processors are allowed
+inline WindowsAffinity get_process_affinity() {
+    HMODULE k32                            = GetModuleHandle(TEXT("Kernel32.dll"));
+    auto    GetThreadSelectedCpuSetMasks_f = GetThreadSelectedCpuSetMasks_t(
+      (void (*)()) GetProcAddress(k32, "GetThreadSelectedCpuSetMasks"));
+
+    WindowsAffinity affinity;
+
+    if (GetThreadSelectedCpuSetMasks_f != nullptr)
+    {
+        USHORT RequiredMaskCount;
+        BOOL   status =
+          GetThreadSelectedCpuSetMasks_f(GetCurrentThread(), nullptr, 0, &RequiredMaskCount);
+
+        // If RequiredMaskCount then these affinities were never set, but it's not consistent
+        // so GetProcessAffinityMask may still return some affinity.
+        if (status != 0)
+        {
+            if (RequiredMaskCount > 0)
+            {
+                std::set<CpuIndex> cpus;
+
+                auto groupAffinities = std::make_unique<GROUP_AFFINITY[]>(RequiredMaskCount);
+
+                GetThreadSelectedCpuSetMasks_f(GetCurrentThread(), groupAffinities.get(),
+                                               RequiredMaskCount, &RequiredMaskCount);
+
+                for (USHORT i = 0; i < RequiredMaskCount; ++i)
+                {
+                    const size_t procGroupIndex = groupAffinities[i].Group;
+
+                    for (size_t j = 0; j < WIN_PROCESSOR_GROUP_SIZE; ++j)
+                    {
+                        if (groupAffinities[i].Mask & (KAFFINITY(1) << j))
+                            cpus.insert(procGroupIndex * WIN_PROCESSOR_GROUP_SIZE + j);
+                    }
+                }
+
+                affinity.newApi = std::move(cpus);
+            }
+        }
+        else
+        {
+            affinity.isDeterminate = false;
+        }
+    }
+
+    DWORD_PTR proc, sys;
+    BOOL      status = GetProcessAffinityMask(GetCurrentProcess(), &proc, &sys);
+    // If proc == 0 then we can't determine affinity because it spans processor groups.
+    if (status != 0 && proc != 0)
```

Comment:
(some personal code style) for some big blocks like these it'd be a nicer to have some early returns instead of nesting multiple levels deep..

Reply:
would need an immediately invoked function expression, not sure if worth it in this particular case, the nesting is not too bad yet

Reply:
well not entirely, the second part of the function can be rewritten like the following.. Also makes me think why we even have these two ways in the same function ? To me it looks like the two parts are mutually exclusive? At least they were in the old code?
https://github.com/official-stockfish/Stockfish/compare/master...Sopel97:Stockfish:numadumbgui#diff-d3deced7cf7b187a18c7c5cac6a88895f2e10502ff710ac4d51e6cbd555d0d04L759

```cpp
    // ...

    DWORD_PTR proc, sys;
    BOOL      status = GetProcessAffinityMask(GetCurrentProcess(), &proc, &sys);

    // If proc == 0 then we can't determine affinity because it spans processor groups.
    if (status == 0 || proc == 0)
    {
        affinity.isDeterminate = false;
        return affinity;
    }

    // If SetProcessAffinityMask was never called the affinity
    // must span all processor groups, but if it was called it must only span one.
    auto [status2, groupAffinity] = get_process_group_affinity();

    if (status2 == 0)
    {
        affinity.isDeterminate = false;
        return affinity;
    }

    if (groupAffinity.size() != 1)
    {
        // It must be the default affinity.
        // We don't set the affinity at all, it will remain std::nullopt
        return affinity;
    }

    std::set<CpuIndex> cpus;
    const size_t       procGroupIndex = groupAffinity[0];
    uint64_t           mask           = static_cast<uint64_t>(proc);

    for (size_t j = 0; j < WIN_PROCESSOR_GROUP_SIZE; ++j)
    {
        if (mask & (KAFFINITY(1) << j))
            cpus.insert(procGroupIndex * WIN_PROCESSOR_GROUP_SIZE + j);
    }

    affinity.oldApi = std::move(cpus);
    return affinity;
```

Reply:
the control flow is not as clear here imo, but I'll try to find some middle ground

> Also makes me think why we even have these two ways in the same function ? To me it looks like the two parts are mutually exclusive? At least they were in the old code?

since we can't determine affinity in all cases (as visible by `isDeterminate` now) the behaviour before this patch was not complete, because I didn't really consider it that important. Now it's as complete as possible.

Reply:
just some personal preference, you can keep that way if you like